src/components/graph/graph.helper.js

```diff
@@ -517,39 +610,39 @@ function getNormalizedNodeCoordinates(
+  let sourceNodeSize = sourceNode?.size || config.node.size;
+  let sourceVectorLength = 1;
 
-      // because this is a circle and A = pi * r^2
-      // we multiply by 0.95, because if we don't the link is not melting properly
-      sourceNodeSize = Math.sqrt(sourceNodeSize / Math.PI) * 0.95;
-
-      // points from the sourceCoords, we move them not to begin in the circle but outside
-      x1 += sourceNodeSize * directionVector.x;
-      y1 += sourceNodeSize * directionVector.y;
-      break;
-    }
+  if (sourceSymbolType === CONST.SYMBOLS.CIRCLE) {
+    sourceVectorLength = calcCircleVectorLength(sourceNodeSize, isSourceCustomNode);
+  } else if (sourceSymbolType === CONST.SYMBOLS.SQUARE) {
+    sourceVectorLength = calcSquareVectorLength(sourceNodeSize, { x: x1, y: y1 }, directionVector, isSourceCustomNode);
+  } else if (sourceSymbolType === CONST.SYMBOLS.RECTANGLE && sourceNodeSize.width && sourceNodeSize.height) {
+    sourceVectorLength = calcRectangleVectorLength(sourceNodeSize, { x: x1, y: y1 }, directionVector);
   }
 
-  switch (targetSymbolType) {
-    case CONST.SYMBOLS.CIRCLE: {
-      // it's fine `markerWidth` or `markerHeight` we just want to fallback to a number
-      // to avoid NaN on `Math.min(undefined, undefined) > NaN
-      let strokeSize = strokeWidth * Math.min(config.link?.markerWidth || 0, config.link?.markerHeight || 0);
-      let targetNodeSize = targetNode?.size || config.node.size;
-
-      // because this is a circle and A = pi * r^2
-      // we multiply by 0.95, because if we don't the link is not melting properly
-      targetNodeSize = Math.sqrt(targetNodeSize / Math.PI) * 0.95;
-
-      // points from the targetCoords, we move the by the size of the radius of the circle + the size of the arrow
-      x2 -= (targetNodeSize + (config.directed ? strokeSize : 0)) * directionVector.x;
-      y2 -= (targetNodeSize + (config.directed ? strokeSize : 0)) * directionVector.y;
-      break;
-    }
+  x1 += sourceVectorLength * directionVector.x;
+  y1 += sourceVectorLength * directionVector.y;
+
+  // it's fine `markerWidth` or `markerHeight` we just want to fallback to a number
+  // to avoid NaN on `Math.min(undefined, undefined) > NaN
+  let strokeSize = strokeWidth * Math.min(config.link?.markerWidth || 5, config.link?.markerHeight || 5);
+  const isTargetCustomNode = targetNode.viewGenerator || config.node.viewGenerator;
+  let targetNodeSize = targetNode?.size || config.node.size;
+  let targetVectorLength = 1;
+
+  if (targetSymbolType === CONST.SYMBOLS.CIRCLE) {
+    targetVectorLength = calcCircleVectorLength(targetNodeSize, isTargetCustomNode);
+  } else if (targetSymbolType === CONST.SYMBOLS.SQUARE) {
+    targetVectorLength = calcSquareVectorLength(targetNodeSize, { x: x2, y: y2 }, directionVector, isTargetCustomNode);
+  } else if (targetSymbolType === CONST.SYMBOLS.RECTANGLE && targetNodeSize.width && targetNodeSize.height) {
```

Comment:
targetNodeSize can be undefined

Reply:
You can use optional chaining in this repo.

```suggestion
  } else if (targetSymbolType === CONST.SYMBOLS.RECTANGLE && targetNodeSize?.width && targetNodeSize?.height) {
```